File: EmojibleIOS/EmojibleIOS/DataSources/ProgramDataSource.swift
//
//  ProgramDataSource.swift
//  EmojibleIOS
//
//  Created by Furkan Yakal on 2.12.2020.
//

import Foundation
import Firebase

class ProgramDataSource {
    
    let database = Firestore.firestore()
    var snapshotListener:ListenerRegistration?
    let notificationCenter = NotificationCenter.default
    var programs: [CodeModel] = []
    var programDataSourceIndices: [String] = []
    var lastPrevEditIndex = 0
    
    func startObservingProgram(){
        guard let currentUser = AuthenticationManager.getInstance().currentUser else { return }
        let userDocRef = database.collection("Users").document(currentUser.uid)
        let programsCollectionRef = userDocRef.collection("Programs")
        snapshotListener = programsCollectionRef.addSnapshotListener{ [unowned self] querySnapshot, error in
            guard let documents = querySnapshot?.documents else { return }
            programs = documents.map{CodeModel(dictionary: $0.data())}
            programDataSourceIndices = documents.map{$0.documentID}
            notificationCenter.post(name: .programsChanged, object: nil, userInfo:["programsChanged":programs])
        }
    }
    
    func stopObservingProgram(){
        guard let snapshotListener = snapshotListener else { return }
        snapshotListener.remove()
    }
    
    func editProgram(oldProgram: CodeModel, newProgram:CodeModel){
        guard let currentUser = AuthenticationManager.getInstance().currentUser else { return }

        for (i, document) in self.programs.enumerated(){
            if document == oldProgram{
                lastPrevEditIndex = i
                break
            }
        }

        let uid = currentUser.uid
        database.collection("Users").document(uid).collection("Programs").document(self.programDataSourceIndices[lastPrevEditIndex]).setData(newProgram.dictionary)
    }
    
    func removeProgram(program: CodeModel){
        guard let currentUser = AuthenticationManager.getInstance().currentUser else { return }
        var index: Int?
        for (i, document) in self.programs.enumerated(){
            if document == program{
                index = i
                break
            }
        }
        let uid = currentUser.uid
        database.collection("Users").document(uid).collection("Programs").document(self.programDataSourceIndices[index!]).delete()
    }
    
    func writeProgram(program: CodeModel){
        guard let currentUser = AuthenticationManager.getInstance().currentUser else { return }
        let uid = currentUser.uid
        self.database.collection("Users").document(uid).collection("Programs").addDocument(data:program.dictionary)
    }
    
    private init(){}
    
    private static let instance = ProgramDataSource()
    public static func getInstance() -> ProgramDataSource{
        return .instance
    
    }
}

File: EmojibleAndroid/app/src/main/java/com/dji/emojibleandroid/dataSources/ProgramDataSource.kt
package com.dji.emojibleandroid.dataSources

import com.dji.emojibleandroid.models.CodeModel
import com.dji.emojibleandroid.services.AuthenticationManager
import com.dji.emojibleandroid.services.Changes
import com.dji.emojibleandroid.services.NotificationCenter
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase

class ProgramDataSource {
    var database = Firebase.firestore
    var snapShotListener: ListenerRegistration? = null
    val notificationCenter: NotificationCenter = NotificationCenter.instance
    var programs: List<CodeModel> = listOf()
    var programDataSourceIndices: List<String> = listOf()

    fun startObservingProgram() {
        val currentUser = AuthenticationManager.instance.currentUser
        val userDocRef = currentUser?.uid?.let { database.collection("Users").document(it) }
        val programsCollectionRef = userDocRef?.collection("Programs")
        snapShotListener = programsCollectionRef?.addSnapshotListener { querySnapshot, _ ->
            val documents = querySnapshot?.documents
            programs = documents?.map { CodeModel(it.data as HashMap<String, Any>) }!!
            programDataSourceIndices = documents.map { it.id }
            notificationCenter.post(Changes.programsChanged, null, "programs" to programs)
        }
    }

    fun stopObservingProgram() {
        snapShotListener?.remove()
    }

    fun editProgram(oldProgram: CodeModel, newProgram: CodeModel) {
        val currentUser = AuthenticationManager.instance.currentUser
        val pModel = programs.find { it == oldProgram } ?: return
        val index = programs.indexOf(pModel)
        val uid = currentUser?.uid
        uid?.let {
            database.collection("Users").document(it).collection("Programs")
                .document(programDataSourceIndices[index!!]).set(newProgram.dictionary)
        }
    }

    fun removeProgram(program: CodeModel) {
        val currentUser = AuthenticationManager.instance.currentUser
        val pModel = programs.find { it == program } ?: return
        val index = programs.indexOf(pModel)
        val uid = currentUser?.uid
        uid?.let {
            database.collection("Users").document(it).collection("Programs")
                .document(programDataSourceIndices[index]).delete()
        }
    }

    fun writeProgram(program: CodeModel) {
        val currentUser = AuthenticationManager.instance.currentUser
        val uid = currentUser?.uid
        uid?.let {
            database.collection("Users").document(it).collection("Programs")
                .add(program.dictionary)
        }
    }

    private object HOLDER {
        val INSTANCE = ProgramDataSource()
    }

    companion object {
        val instance: ProgramDataSource by lazy { HOLDER.INSTANCE }
    }
}